nemo/core/neural_graph/neural_graph.py

```diff
@@ -0,0 +1,882 @@
+            # Find the tensors within this (outer) graph that are outputs by the same producer-port.
+            bumped_step = tensor.producer_step_number + step_bump
+            # producer_name = tensor.producer_name
+            producer_port_name = tensor.name
+            # Get adequate tensor from "outer graph" (self).
+            output_tensors[key] = self.tensors[bumped_step][producer_port_name]
+
+        if len(output_tensors) == 1:
+            # Return a single tensor.
+            key = list(output_tensors)[0]
+            results = output_tensors[key]
+
+            # Bind the "default" output ports of the inner graph as "default" output ports of this graph.
+            # Call the bind() method of bound_outputs directly, as we already have the tensors in our graph.
+            # But: Use output port name of the inner graph!
+            self.outputs.bind([results], [key])
+
+        else:
+            # Create a named tuple type enabling to access outputs by attributes (e.g. out.x).
+            output_class_name = f'{self.__class__.__name__}Output'
+            result_type = namedtuple(typename=output_class_name, field_names=output_tensors.keys())
+
+            # Return the bound output tensors.
+            results = result_type(*output_tensors.values())
+
+            # Bind the "default" output ports of the inner graph as "default" output ports of this graph.
+            # Call the bind() method of bound_outputs directly, as we already have the tensors in our graph.
+            # But: Use output port name of the inner graph!
+            self.outputs.bind(output_tensors.values(), output_tensors.keys())
+
+        # Ok, now we can turn automatic binding on.
+        self.default_output_binding = True
+
+        # Return the results.
+        return results
+
+    def record_step(self, module):
+        """
+            Records the operation (module plus passed inputs) on a list.
+
+            Args:
+                module: Neural modules added to a given graph.
+
+            Returns:
+                Step number.
+        """
+        # The solution allows loops in the graph.
+        # This also means that module with that name can already be present in the graph.
+        if module.name in self._modules.keys():
+            # Check if this is the same module.
+            if self._modules[module.name] is not module:
+                raise KeyError("Neural Graph already contains a different module with name `{}`!".format(module.name))
+
+        else:
+            # Add module to list of modules.
+            self._modules[module.name] = module
+
+        # Add step - store the module name.
+        step_number = len(self._steps)
+        self._steps[step_number] = module.name
+
+        # Return the current step number.
+        return step_number
+
+    @property
+    def step_number(self):
+        """ Returns:
+                Last step number. """
+        return len(self._steps) - 1
```

Comment:
Yes - step numeration starts from 0, so when you have 3 steps in your list, the last (i.e.) current step is 2.

Reply:
For a property called step_number, this is rather unintuitive.
It also seems unintuitive for a len() property to return the last index instead of the len(). Users should just know to add a -1 or do range(0, len(obj))

Reply:
Hmmm, not really understand what are you saying here...

Let's take this example:

The `graph0` Neural Graph:
     * Modules (3):
        * `module0` (RealFunctionDataLayer)
        * `module1` (TaylorNet)
        * `module2` (MSELoss)
     * Steps (3):
        0. module0
        1. module1
        2. module2
     * Connections (3):
        * 0.module0.x->1.module1.x | axes: (batch, dimension);  elements_type: ChannelType
        * 0.module0.y->2.module2.target | axes: (batch, dimension);  elements_type: LabelsType
        * 1.module1.y_pred->2.module2.predictions | axes: (batch, dimension);  elements_type: ChannelType
     * Graph Inputs (0):
     * Graph Outputs (4, default):
        * 0.module0.x->x | axes: (batch, dimension);  elements_type: ChannelType
        * 0.module0.y->y | axes: (batch, dimension);  elements_type: LabelsType
        * 1.module1.y_pred->y_pred | axes: (batch, dimension);  elements_type: ChannelType
        * 2.module2.loss->loss | axes: None;  elements_type: LossType

Steps store the "execution order", i.e. first execute module0, then module1, then module2.

When e.g. "executing" step 2 the property returns: 2. That value is then "injected" into all tensors produced as a result of that execution. 

Not really sure what are the concerns here... the name of that property?

Reply:
I expect that a property that return the length of such an object to return 3 and not 2.

Reply:
"I expect that a property that return the length" ?

Please note that NG enables to access modules using their names, like a dictionary, e.g.
graph["my_module"] - > returns my_module NM.
This is one of the API designs that we have "formalized" in the 'Module Composition" specification.
As such, graph has a **length** (__len__(self) method) and len(graph) returns number of NMs in it.

step_number is "number of the actually executed step". It has nothing to do with length.

Reply:
If you want to keep it as `len(object) - 1`, then mark this property as private.
It makes no sense, why an API would return `len(object) - 1` instead of `len(object)`